lib/utils/resolve-node-module.ts

```diff
@@ -0,0 +1,148 @@
+              `./${remainingPath}`
+            ] as string
+            const fullExportPath = `${nodeModulesPath}/${exportPath.replace(/^\.\//, "")}`
+
+            // Check if the export path exists
+            if (fsMap[fullExportPath]) return fullExportPath
+
+            // Try with extensions
+            for (const ext of extensions) {
+              const pathWithExt =
+                fullExportPath.replace(/\.js$|\.jsx$/, "") + ext
+              if (fsMap[pathWithExt]) return pathWithExt
+            }
+          }
+
+          // Handle conditional exports
+          if (
+            remainingPath === "" &&
+            packageJson.exports["import"] &&
+            typeof packageJson.exports["import"] === "string"
+          ) {
+            const exportPath = packageJson.exports["import"] as string
+            const fullExportPath = `${nodeModulesPath}/${exportPath.replace(/^\.\//, "")}`
+
+            // Check if the export path exists
+            if (fsMap[fullExportPath]) return fullExportPath
+
+            // Try with extensions
+            for (const ext of extensions) {
+              const pathWithExt =
+                fullExportPath.replace(/\.js$|\.jsx$/, "") + ext
+              if (fsMap[pathWithExt]) return pathWithExt
+            }
+          }
+        }
+
+        // Fall back to module/main fields if exports didn't resolve
+        const entryPoint = packageJson.module || packageJson.main || "index.js"
+        const fullPath = `${nodeModulesPath}/${entryPoint}`
+
+        // Check if the entry point exists
+        if (fsMap[fullPath]) return fullPath
+
+        // Try with extensions
+        for (const ext of extensions) {
+          const pathWithExt = fullPath.replace(/\.js$|\.jsx$/, "") + ext
+          if (fsMap[pathWithExt]) return pathWithExt
+        }
+      } catch (error) {
+        // Silently continue if package.json parsing fails
+      }
+    }
+
+    // If there's a remaining path, try to resolve it
+    if (remainingPath) {
+      const fullPath = `${nodeModulesPath}/${remainingPath}`
+      if (fsMap[fullPath]) return fullPath
+
+      // Try with extensions
+      for (const ext of extensions) {
+        const pathWithExt = fullPath + ext
+        if (fsMap[pathWithExt]) return pathWithExt
+      }
+
+      // Try index files
+      for (const ext of extensions) {
+        const indexPath = `${fullPath}/index${ext}`
+        if (fsMap[indexPath]) return indexPath
+      }
+    } else {
+      // Try index files in the module root
+      for (const ext of extensions) {
+        const indexPath = `${nodeModulesPath}/index${ext}`
+        if (fsMap[indexPath]) return indexPath
+      }
+    }
+
+    // If not found and we have a parent directory, try there
+    const parentPath = dirname(searchPath)
+    if (parentPath && parentPath !== searchPath) {
+      return tryResolveInNodeModules(parentPath)
+    }
+
+    return null
+  }
+
+  return tryResolveInNodeModules(basePath)
+}
```

Comment:
The code structure here uses a recursive inline function, instead of that, you should use a `Context` pattern

recursive inline functions are notoriously impossible to refactor. This file will just grow and grow. If you define a nice `Context` interface, you'll find that it's much much easier to split this up into smaller functions

Reply:
Here's an example of a `Context`:

```tsx
interface NodeResolutionContext {
  fsMap: Record<string, string>
}
```

You can then have methods like:

```tsx
export const findPackageJson = (path: string, ctx: NodeResolutionContext) => {
  // ...
}
```